packages/waallet/content/provider.ts

```diff
@@ -1,22 +1,35 @@
 import { EventEmitter } from "events"
 
 import { type BackgroundMessenger } from "~packages/messenger/background"
+import type { JsonRpcResponse } from "~packages/rpc/json/provider"
 import { format } from "~packages/util/json"
 import type { WebAuthnCreation, WebAuthnRequest } from "~packages/webAuthn"
 
 import { WaalletMessage } from "../message"
 import { type WaalletRequestArguments } from "../rpc"
 
+type WalletProviderError = Error & {
+  code?: number
+  data?: unknown
+}
 export class WaalletContentProvider extends EventEmitter {
   public constructor(private backgroundMessenger: BackgroundMessenger) {
     super()
   }
 
   public async request<T>(args: WaalletRequestArguments): Promise<T> {
-    const res = await this.backgroundMessenger.send({
+    const res: JsonRpcResponse<T> = await this.backgroundMessenger.send({
       name: WaalletMessage.JsonRpcRequest,
       body: args
     })
+
+    // https://github.com/ethers-io/ethers.js/blob/72c2182d01afa855d131e82635dca3da063cfb31/src.ts/providers/provider-browser.ts#L69-L85
+    if (res.error) {
+      const error: WalletProviderError = new Error(res.error.message)
+      error.code = res.error.code
+      error.data = res.error.data
+      throw error
+    }
```

Comment:
Maybe we could reuse `JsonRpcError` here?

Reply:
The `JsonRpcError` format(code and data are wrapped inside) doesn't seem to be what `ethers` expects. 

-  `JsonRpcError` output:
    ```
      {
      "name": "JsonRpcError",
      "jsonrpc": "2.0",
      "id": 0,
      "error": {
        "code": -32521,
        "message": "execution reverted",
        "data": {
          "revertData": "0x"
        }
      }
    ```

- [Ethers](https://github.com/ethers-io/ethers.js/blob/72c2182d01afa855d131e82635dca3da063cfb31/src.ts/providers/provider-browser.ts#L69-L85):
  ```
  try {...} 
  catch (e: any) {
         const error = new Error(e.message);
         (<any>error).code = e.code;
         (<any>error).data = e.data;
         (<any>error).payload = payload;
         this.emit("debug", { action: "receiveEip1193Error", error });
         throw error;
  }
  ```

Reply:
Maybe we can modify it a bit like your previous attempt

```typescript
export class JsonRpcError extends Error {
  // ...
  public code: number
  public message: string
  public data?: unknown
```

Reply:
Changed:
1. rename to `ProviderRpcError` (same as EIP-1193 error interace name)
2. `wrap` function convert `JsonRpcResponse` to `ProviderRpcError` object
3. `unwrap` function convert `ProviderRpcError` object to `JsonRpcResponse`

[re-define error](https://github.com/pilagod/waallet-extension/pull/179/commits/b99fb50f2174ac9c581bbfb2cd3d46aa149051c1)